From 2bcd16a9e250b21f8e7ad8cbf945266213036cde Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Fri, 15 Jan 2016 09:33:18 +0800
Subject: [PATCH 1426/1691] MLK-12263 ARM: imx: disable arm_arch_timer for
 i.MX7D nosmp

commit 2bcd16a9e250b21f8e7ad8cbf945266213036cde from
git://git.freescale.com/imx/linux-2.6-imx.git imx_4.1.15_1.0.0_ga

When SMP is deselected, ARM_ARCH_TIMER is still enabled while
broadcast time is disabled, so when system enters WAIT mode,
ARM platform's clock will be disabled, then system tick timer
will stop and cause system stay at WAIT mode and timer event
will NOT come as expected.

To fix this issue, we do runtime check in kernel boot up,
if SMP is NOT enabled, ARM_ARCH_TIMER will be disabled and
using GPT timer always.

we have to put this check in early stage before common
arm_arch_timer driver probed.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
---
 arch/arm/mach-imx/mach-imx7d.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/arch/arm/mach-imx/mach-imx7d.c b/arch/arm/mach-imx/mach-imx7d.c
index 101c140..8c5b71e 100644
--- a/arch/arm/mach-imx/mach-imx7d.c
+++ b/arch/arm/mach-imx/mach-imx7d.c
@@ -19,6 +19,12 @@
 #include "common.h"
 #include "cpuidle.h"
 
+static struct property device_disabled = {
+	.name = "status",
+	.length = sizeof("disabled"),
+	.value = "disabled",
+};
+
 static int ar8031_phy_fixup(struct phy_device *dev)
 {
 	u16 val;
@@ -108,6 +114,17 @@ static inline void imx7d_enet_init(void)
 	imx7d_enet_clk_sel();
 }
 
+static inline void imx7d_disable_arm_arch_timer(void)
+{
+	struct device_node *node;
+
+	node = of_find_compatible_node(NULL, NULL, "arm,armv7-timer");
+	if (node) {
+		pr_info("disable arm arch timer for nosmp!\n");
+		of_add_property(node, &device_disabled);
+	}
+}
+
 static void __init imx7d_init_machine(void)
 {
 	struct device *parent;
@@ -128,6 +145,9 @@ static void __init imx7d_init_irq(void)
 	imx_init_revision_from_anatop();
 	imx_src_init();
 	irqchip_init();
+#ifndef CONFIG_SMP
+	imx7d_disable_arm_arch_timer();
+#endif
 }
 
 static void __init imx7d_init_late(void)
-- 
1.9.1

